src/CoreBundle/Tests/CronJob/CronJobSchedulerTest.php

```diff
@@ -0,0 +1,316 @@
+use ChameleonSystem\CoreBundle\Interfaces\ChameleonTimeProviderInterface;
+use PHPUnit\Framework\TestCase;
+use Prophecy\Argument;
+use Prophecy\Prophecy\ObjectProphecy;
+use Symfony\Component\Validator\Constraints\DateTime;
+
+class CronJobSchedulerTest extends TestCase
+{
+    /**
+     * @var DateTime|null
+     */
+    private $mockCurrentUtcTime;
+    /**
+     * @var \DateTimeZone
+     */
+    private $mockTimeZone;
+    /**
+     * @var ObjectProphecy|ChameleonTimeProviderInterface
+     */
+    private $mockTimeProvider;
+
+    /**
+     * @var CronJobScheduler
+     */
+    private $subject;
+    /**
+     * @var mixed
+     */
+    private $subjectResult;
+
+    public function provideDataForExecutionNeeded(): array
+    {
+        $schedule24Hours = new CronJobScheduleDataModel(
+            1440,
+            2880,
+            false,
+            $this->createDate('2019-11-14 01:15:00')
+        );
+
+        return [
+            'requires-execution-at-midnight' => [
+                'schedule' => new CronJobScheduleDataModel(
+                    10,
+                    2880,
+                    false,
+                    $this->createDate('2019-11-14 23:15:00')
+                ),
+                'currentUtcTime' => $this->createDate('2019-11-15 00:00:00'),
+                'expectedResult' => true,
+            ],
+
+            'requires-execution-exact-24-hours-ago' => [
+                'schedule' => $schedule24Hours,
+                'currentUtcTime' => $this->createDate('2019-11-15 01:15:00'),
+                'expectedResult' => true,
+            ],
+            'requires-execution-25-hours-ago' => [
+                'schedule' => $schedule24Hours,
+                'currentUtcTime' => $this->createDate('2019-11-15 02:15:00'),
+                'expectedResult' => true,
+            ],
+            'no-execution-23-hours-ago' => [
+                'schedule' => $schedule24Hours,
+                'currentUtcTime' => $this->createDate('2019-11-15 00:15:00'),
+                'expectedResult' => false,
+            ],
+            'no-execution-one-second-before-next-planned-job' => [
+                'schedule' => $schedule24Hours,
+                'currentUtcTime' => $this->createDate('2019-11-15 01:14:59'),
+                'expectedResult' => false,
+            ],
+            'requires-execution-but-is-locked' => [
+                'schedule' => new CronJobScheduleDataModel(
+                    1440,
+                    2880,
+                    true,
+                    $this->createDate('2019-11-14 01:15:00')
+                ),
+                'currentUtcTime' => $this->createDate('2019-11-15 02:15:00'),
+                'expectedResult' => false,
+            ],
+            'requires-execution-is-locked-but-lock-timed-out-for-30-min' => [
+                'schedule' => new CronJobScheduleDataModel(
+                    1440,
+                    1470,
```

Comment:
This should be 30?

Reply:
no. Lock time is evaluated from the time the cron job started - but only after the executeEveryNMinutes time has been passed (there is no point unlocking the job when it is not required to run).

I added a test case to check that the lock is ignored when the job is not scheduled for execution

Reply:
but it is the "unlockAfterNMinutes" field  and 1470 does not really make sense for that field (?)